src/Workspaces/Core/Portable/Rename/Renamer.SyncNamespaceDocumentAction.cs

```diff
@@ -0,0 +1,118 @@
+        {
+            private readonly AnalysisResult _analysis;
+
+            private SyncNamespaceDocumentAction(AnalysisResult analysis, OptionSet optionSet, ImmutableArray<ErrorResource> errors)
+                : base(errors, optionSet)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+             => WorkspacesResources.ResourceManager.GetString("Sync_namespace_to_folder_structure", culture ?? WorkspacesResources.Culture)!;
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Document document, CancellationToken cancellationToken)
+            {
+                var solution = document.Project.Solution;
+
+                if (!_analysis.SupportsSyncNamespace)
+                {
+                    return solution;
+                }
+
+                var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
+
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Don't descend into anything other than top level declarations from the root.
+                // ChangeNamespaceService only controls top level declarations right now
+                var namespaceDeclarations = syntaxRoot.DescendantNodes(n => !syntaxFacts.IsDeclaration(n)).Where(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+                foreach (var namespaceDeclaration in namespaceDeclarations)
+                {
+                    solution = await changeNamespaceService.ChangeNamespaceAsync(
+                        document,
+                        namespaceDeclaration,
+                        targetNamespace,
+                        cancellationToken).ConfigureAwait(false);
+
+                    document = solution.GetRequiredDocument(document.Id);
+                }
+
+                return solution;
+
+                static string BuildNamespaceFromFolders(
+                    IEnumerable<string> folders,
+                    ISyntaxFacts syntaxFacts)
+                {
+                    var parts = folders.SelectMany(folder => folder.Split(new[] { '.' }));
+
+                    // string.Empty is used to represent the global namespace
+                    // when calling ChangeNamespaceAsync
+                    return parts.All(syntaxFacts.IsValidIdentifier) ? string.Join(".", parts) : string.Empty;
+                }
+            }
+
+            public static Task<SyncNamespaceDocumentAction?> TryCreateAsync(Document document, IReadOnlyList<string> newFolders, OptionSet optionSet, CancellationToken _)
+            {
+                var errors = new ArrayBuilder<ErrorResource>();
```

Comment:
```suggestion
                using var _ = ArrayBuilder<ErrorResource>.GetInstance(out var errors);
```
(update below to be ToImmutable as well).